Tidy-ups to gen_relocate_app.py

jhol opened the pull request:
* Apply Ruff code formatting
  * Various code conciseness improvements
  * Use named fields in string template to improve clarity
  * Use `bool` when applicable instead of `int`
  * Fix all ruff linting warnings with the exception of E101.

jhol committed:
code_relocation: combine typing imports onto a single line

The script can be made more concise by combining the imports of three
classes from the typing module into a single line.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: use named fields in string templates

Previously the script used templates where the fields were identified by
numeric ordinal-identified placeholders. This makes the code hard to read
because it is hard to tell which string corresponds to which placeholder
in the substitution.

This patch corrects the issue by replacing the ordinal placeholders with
named placeholders.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: improve code conciseness

This patch makes various simplifications to the code structure which make
it more concise by reducing repetition.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: use booleans for string_create_helper arguments

The string_create_helper function has the arguments: load_address_in_flash
and is_copy. These arguments take logical values, but previously the
calling code used 1 and 0 rather than the more idiomatic True and False.

This patch corrects the issue by replacing use of int values with bool
values.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

JarmouniA reviewed (dismissed):
Looks quite nice, thanks!

jhol commented:
> @jhol There are several CI failures that need attention, related to templates' named args.

I think it should be fixed now.

tejlmand commented on scripts/build/gen_relocate_app.py:
```diff
@@ -226,7 +226,9 @@ def region_is_default_ram(region_name: str) -> bool:
 
 
-def find_sections(filename: str, symbol_filter: str) -> 'dict[SectionKind, list[OutputSection]]':
+def find_sections(
+    filename: str, symbol_filter: str
+) -> "dict[SectionKind, list[OutputSection]]":
```
Zephyr allows 100 char line length, so why this line wrap ?

tejlmand commented on scripts/build/gen_relocate_app.py:
```diff
@@ -126,48 +126,48 @@ class OutputSection(NamedTuple):
+	}} {load_address}
+        __{memory_type}_{kind}_reloc_end = .;
+        __{memory_type}_{kind}_reloc_start = ADDR(.{memory_type}_{kind}_reloc);
+        __{memory_type}_{kind}_reloc_size = __{memory_type}_{kind}_reloc_end - __{memory_type}_{kind}_reloc_start;
```
when changing from numbers to named placeholders, then please care for line length.

jhol replied:
Fixed by renaming the `{memory_type}` placeholder to `{mem}`.

jhol replied:
Replaced `{memory_type}` with `{mem}` so as to be more concise.

tejlmand reviewed (changes requested):
Thanks for this cleanup, generally a nice improvement.

But before approving changes based on ones favorite tool, then I think such changes should only be done after proper discussion and agreement.

carlescufi commented:
> But before approving changes based on ones favorite tool, then I think such changes should only be done after proper discussion and agreement.

@pdgendt any thoughts on this?

pdgendt commented:
> > But before approving changes based on ones favorite tool, then I think such changes should only be done after proper discussion and agreement.
> 
> @pdgendt any thoughts on this?

Yes, in zephyr we use `ruff`, so please use `ruff format` instead. The configuration is to have max 100 chars line length, and keep quoting characters.

pdgendt commented:
One important note: this file is currently excluded in `zephyr/.ruff-excludes.toml`, so remove the exclusion and fix formatting/issues reported by `ruff`.

tejlmand commented:
> Yes, in zephyr we use `ruff`, so please use `ruff format` instead.

@pdgendt  Thanks, I learned something new today :smile: 

Guess a Python coding guide and ruff tool mentioning on this page would be welcome
https://docs.zephyrproject.org/latest/contribute/style/index.html#coding-style-guidelines

pdgendt commented:
> Guess a Python coding guide and ruff tool mentioning on this page would be welcome [docs.zephyrproject.org/latest/contribute/style/index.html#coding-style-guidelines](https://docs.zephyrproject.org/latest/contribute/style/index.html#coding-style-guidelines)

Yeah, will try to put something together.

jhol committed:
code_relocation: applied ruff code formatting to gen_relocate_app

Ruff is the Zephyr projects supported Python formatting tool. This patch
applies auto-formatting gen_relocate_app.py in preparation for coming
tidy-ups and improvements.

With the Ruff auto-formatter applied, error E501 can be removed from
.ruff-excludes.toml exclusion rules.

gen_relocate_app.py has also been removed from the format exclude list.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: fix ruff B028 warnings

The ruff python linter produces a B028 warning when the warnings.warn()
function is called with a stacklevel parameter.

By default the function will set a stacklevel of 1 which causes it to
output the stack frame of the line where the function is called without
any context information from higher up the stack.

It is recommended to use a stacklevel of 2 or higher. Therfore, this patch
sets the stacklevel parameter of all warnings.warn() calls to to 2.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: fix ruff I001 warning

The ruff python linter produces a I001 warning when the imports of a Python
script are not sorted. This patch corrects the issue by sorting them.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: fix ruff SIM102 warning

The ruff python linter produces a SIM102 warning when there are nested if
statements which can be collapsed into a single if statement:

    if foo:
        if bar:
	    ...

...becomes...

    if foo and bar:
        ...

This patch corrects the single instance of this issue in the script.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

pdgendt commented:
@jhol thanks! Could you remove the entry here, so that CI doesn't exclude it anymore.

https://github.com/zephyrproject-rtos/zephyr/blob/b926ba6abf71880ffd4e4803a9f31cd51539d14d/.ruff-excludes.toml#L1479

jhol committed:
code_relocation: fix ruff SUM401 warning

The ruff python linter produces a SIM401 warning when a dictionary is
accessed using if-statements to check for key presence. In this case, the
dict.get() method could be used instead.

For example:

    value = foo["bar"] if "bar" in foo else 0

...can be replaced with:

    value = foo.get("bar", 0)

This patch corrects the single instance of this issue in the script.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: fix ruff UP006 and UP035 warnings

PEP585 enabled collections in the Python standard library (like tuple) to
be used as generic type annotations directly, instead of importing
analogous members from the typing module (like typing.Tuple).

The ruff python linter produces a UP006 warning if the deprecated type
annotations continue to be used.

This patch corrects the issue by correcting the single instance where
PEP585-style type annotations can be used.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol committed:
code_relocation: fix ruff UP037 warnings

In Python, type annotations can be quoted to avoid forward references.

However, if "from __future__ import annotations" is present, Python will
always evaluate type annotations in a deferred manner, making the quotes
unnecessary.

The ruff python linter produces UP037 warnings to indicate cases where
quoting of type annotations can be removed. This patch corrects the
warnings.

Signed-off-by: Joel Holdsworth <jholdsworth@nvidia.com>

jhol commented:
> @jhol thanks! Could you remove the entry here, so that CI doesn't exclude it anymore.

Done.

I also fixed most of the linter warnings.

tejlmand reviewed (approved):
Thanks for the cleanup.
